source/common/common/matchers.h

```diff
@@ -6,49 +6,99 @@
 #include "envoy/type/matcher/metadata.pb.h"
 #include "envoy/type/matcher/number.pb.h"
 #include "envoy/type/matcher/string.pb.h"
+#include "envoy/type/matcher/value.pb.h"
 
 #include "common/common/utility.h"
-
-#include "absl/types/optional.h"
+#include "common/protobuf/protobuf.h"
 
 namespace Envoy {
 namespace Matchers {
 
-class DoubleMatcher {
+class ValueMatcher;
+typedef std::shared_ptr<const ValueMatcher> MatcherPtr;
+
+class ValueMatcher {
 public:
-  DoubleMatcher(const envoy::type::matcher::DoubleMatcher& matcher) : matcher_(matcher) {}
+  virtual ~ValueMatcher() {}
 
   /**
    * Check whether the value is matched to the matcher.
-   * @param value the double value to check.
-   * @return true if it's matched otherwise false.
    */
-  bool match(double value) const;
+  virtual bool match(const ProtobufWkt::Value& value) const PURE;
+
+  /**
+   * Create the matcher object.
+   */
+  static MatcherPtr create(const envoy::type::matcher::ValueMatcher& value);
+};
+
+class NullMatcher : public ValueMatcher {
+public:
+  NullMatcher() {}
+
+  /**
+   * Check whether the value is NULL.
+   */
+  bool match(const ProtobufWkt::Value& value) const override;
+};
+
+class BoolMatcher : public ValueMatcher {
+public:
+  BoolMatcher(bool matcher) : matcher_(matcher) {}
+
+  bool match(const ProtobufWkt::Value& value) const override;
+
+private:
+  bool matcher_;
+};
+
+class PresentMatcher : public ValueMatcher {
+public:
+  PresentMatcher(bool matcher) : matcher_(matcher) {}
+
+  bool match(const ProtobufWkt::Value& value) const override;
+
+private:
+  bool matcher_;
+};
+
+class DoubleMatcher : public ValueMatcher {
+public:
+  DoubleMatcher(const envoy::type::matcher::DoubleMatcher& matcher) : matcher_(matcher) {}
+
+  bool match(const ProtobufWkt::Value& value) const override;
 
 private:
   const envoy::type::matcher::DoubleMatcher matcher_;
 };
 
-class StringMatcher {
+class StringMatcher : public ValueMatcher {
 public:
   StringMatcher(const envoy::type::matcher::StringMatcher& matcher) : matcher_(matcher) {
     if (matcher.match_pattern_case() == envoy::type::matcher::StringMatcher::kRegex) {
       regex_ = RegexUtil::parseRegex(matcher_.regex());
     }
   }
 
-  /**
-   * Check whether the value is matched to the matcher.
-   * @param value the string to check.
-   * @return true if it's matched otherwise false.
-   */
-  bool match(const std::string& value) const;
+  bool match(const ProtobufWkt::Value& value) const override;
```

Comment:
I don't think the comments here is needed. It's overriding match() defined in parent class ValueMatcher. If we add the comment back, it would just repeats the one in ValueMatcher class.